src/gt4py/next/embedded/common.py

```diff
@@ -0,0 +1,144 @@
+            new_index = (rng.start if idx >= 0 else rng.stop) + idx
+            if new_index < rng.start or new_index >= rng.stop:
+                raise embedded_exceptions.IndexOutOfBounds(
+                    domain=domain, indices=index, index=idx, dim=dim
+                )
+
+    return common.Domain(*named_ranges)
+
+
+def _absolute_sub_domain(domain: common.Domain, index: common.DomainSlice) -> common.Domain:
+    named_ranges: list[common.NamedRange] = []
+    for i, (dim, rng) in enumerate(domain):
+        if (pos := _find_index_of_dim(dim, index)) is not None:
+            named_idx = index[pos]
+            idx = named_idx[1]
+            if isinstance(idx, common.UnitRange):
+                if not idx <= rng:
+                    raise embedded_exceptions.IndexOutOfBounds(
+                        domain=domain, indices=index, index=named_idx, dim=dim
+                    )
+
+                named_ranges.append((dim, idx))
+            else:
+                # not in new domain
+                assert common.is_int_index(idx)
+                if idx < rng.start or idx >= rng.stop:
+                    raise embedded_exceptions.IndexOutOfBounds(
+                        domain=domain, indices=index, index=named_idx, dim=dim
+                    )
+        else:
+            # dimension not mentioned in slice
+            named_ranges.append((dim, domain.ranges[i]))
+
+    return common.Domain(*named_ranges)
+
+
+def _tuplize_field_slice(v: common.FieldSlice) -> common.FieldSlice:
+    """
+    Wrap a single index/slice/range into a tuple.
+
+    Note: the condition is complex as `NamedRange`, `NamedIndex` are implemented as `tuple`.
+    """
+    if (
+        not isinstance(v, tuple)
+        and not common.is_domain_slice(v)
+        or common.is_named_index(v)
+        or common.is_named_range(v)
+    ):
+        return cast(common.FieldSlice, (v,))
+    return v
+
+
+def _expand_ellipsis(
+    indices: tuple[common.IntIndex | slice | EllipsisType, ...], target_size: int
+) -> tuple[common.IntIndex | slice, ...]:
+    expanded_indices: list[common.IntIndex | slice] = []
+    for idx in indices:
+        if idx is Ellipsis:
+            expanded_indices.extend([slice(None)] * (target_size - (len(indices) - 1)))
+        else:
+            expanded_indices.append(idx)
+    return tuple(expanded_indices)
+
+
+def _slice_range(input_range: common.UnitRange, slice_obj: slice) -> common.UnitRange:
+    if slice_obj == slice(None):
+        return common.UnitRange(input_range.start, input_range.stop)
+
+    start = (
+        input_range.start if slice_obj.start is None or slice_obj.start >= 0 else input_range.stop
+    ) + (slice_obj.start or 0)
+    stop = (
+        input_range.start if slice_obj.stop is None or slice_obj.stop >= 0 else input_range.stop
+    ) + (slice_obj.stop or len(input_range))
+
+    if start < input_range.start or stop > input_range.stop:
+        raise IndexError()
```

Comment:
Missing exception message

Reply:
The only purpose is to catch the exception at the only call site. I don't like the pattern, but nobody should ever see the exception.

Reply:
I would add a message anyway, even it's basically just a comment.

Reply:
done